cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,304 @@
+		container.Events = append(container.Events, *event)
+		container.LastEventTime = &(*event).Time
+		container.Liveliness = metal.MachineLivelinessAlive
+	}
+}
+
+func handlePlannedRebootEvent(e *fsm.Event) {
+	container, _, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	container.IncompleteCycles = 0
+	container.Liveliness = metal.MachineLivelinessAlive
+}
+
+func handlePhonedHomeEvent(e *fsm.Event) {
+	container, event, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	if e.Src == string(metal.ProvisioningEventPlannedReboot) {
+		if event.Time.Sub(*container.LastEventTime) > timeOutAfterPlannedReboot {
+			container.Liveliness = metal.MachineLivelinessUnknown
+		}
+	} else {
+		container.Events = append(container.Events, *event)
+		container.LastEventTime = &(*event).Time
+		container.Liveliness = metal.MachineLivelinessAlive
+	}
+
+	container.IncompleteCycles = 0
+}
+
+func handleAliveEvent(e *fsm.Event) {
+	container, _, log, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	log.Debugw("received provisioning alive event", "id", container.ID)
+	container.Liveliness = metal.MachineLivelinessAlive
+}
+
+func parseCallbackArguments(args []interface{}) (*EventContainer, *metal.ProvisioningEvent, *zap.SugaredLogger, error) {
+	if len(args) != 3 {
+		return nil, nil, nil, errors.New("expecting arguments of types *EventContainer, *metal.ProvisioningEvent, *zap.SugaredLogger")
+	}
+
+	container, ok := args[0].(*EventContainer)
+	if !ok {
+		return nil, nil, nil, errors.New("first argument must be of type *EventContainer")
+	}
+
+	event, ok := args[1].(*metal.ProvisioningEvent)
+	if !ok {
+		return nil, nil, nil, errors.New("second argument must be of type *metal.ProvisioningEvent")
+	}
+
+	log, ok := args[2].(*zap.SugaredLogger)
+	if !ok {
+		return nil, nil, nil, errors.New("third argument must be of type *zap.SugaredLogger")
+	}
+
+	return container, event, log, nil
+}
+
+func newProvisioningFSM(initialState metal.ProvisioningEventType) *fsm.FSM {
+	provisioningFSM := fsm.NewFSM(
+		string(initialState),
+		events,
+		callbacks,
+	)
+
+	return provisioningFSM
+}
+
+// HandleProvisioningEvent writes the ProvisioningEvent to the the ProvisioningEventContainer and checks if the events up to this point
+// occured in the expected order
+func HandleProvisioningEvent(event *metal.ProvisioningEvent, container *metal.ProvisioningEventContainer, log *zap.SugaredLogger) (*metal.ProvisioningEventContainer, error) {
+	provisioningFSM := newProvisioningFSM(container.Events[0].Event)
+	eventContainer := EventContainer{
+		IncompleteCycles: 0,
+		ID:               container.ID,
+		Liveliness:       container.Liveliness,
+		Events:           []metal.ProvisioningEvent{container.Events[0]},
+		LastEventTime:    container.LastEventTime,
+	}
+
+	var err error
+	var invalidEventError fsm.InvalidEventError
+	for i, e := range container.Events {
```

Comment:
When you use just the flag `Crashloop` there is no need to cycle through the event history anymore.